examples/getitem/src/lib.rs

```diff
@@ -0,0 +1,64 @@
+// This is a very fake example of how to check __getitem__ parameter and handle appropriately
+use pyo3::exceptions::PyTypeError;
+use pyo3::prelude::*;
+use pyo3::types::PySlice;
+use std::os::raw::c_long;
+
+#[pyclass]
+struct ExampleContainer {
+    // represent the maximum length our container is pretending to be
+    max_length: i32,
+}
+
+#[pymethods]
+impl ExampleContainer {
+    #[new]
+    fn new() -> Self {
+        ExampleContainer { max_length: 100 }
+    }
+
+    fn __getitem__(&self, key: &PyAny) -> PyResult<i32> {
```

Comment:
As a passing thought, it would be quite nice to eventually put in this example  `key: IntOrSlice<'py>` where that type is defined as

```rust
#[derive(FromPyObject)]
enum IntOrSlice<'py> {
    Int(i32),
    Slice(&'py PySlice),
}
```

I think the only reason that you wouldn't want to consider that today is performance of the error cases, if I recall correctly?

Reply:
I'm not familiar with the performance delta, but I added this enum possibility on the __setitem__ side to show that methodology as well